From 8d73da5c090e8ddac74063021549e796715e2def Mon Sep 17 00:00:00 2001
From: Hien Huynh <hien.huynh.px@renesas.com>
Date: Tue, 27 Oct 2020 14:47:04 +0700
Subject: [PATCH 14/18] rzg: drivers: SPI multi I/O bus controller

This commit adds support SPI multi I/O bus controller to control
flash memory in rzg gen2 platform.
Refer from commit "rcar_gen3: drivers: spi multio bus controller".

Signed-off-by: Hien Huynh <hien.huynh.px@renesas.com>
---
 drivers/renesas/rzg/rpc/rpc_driver.c    | 57 +++++++++++++++++++++++++++++++++
 drivers/renesas/rzg/rpc/rpc_registers.h | 25 +++++++++++++++
 2 files changed, 82 insertions(+)
 create mode 100644 drivers/renesas/rzg/rpc/rpc_driver.c
 create mode 100644 drivers/renesas/rzg/rpc/rpc_registers.h

diff --git a/drivers/renesas/rzg/rpc/rpc_driver.c b/drivers/renesas/rzg/rpc/rpc_driver.c
new file mode 100644
index 0000000..ffa9f4f
--- /dev/null
+++ b/drivers/renesas/rzg/rpc/rpc_driver.c
@@ -0,0 +1,57 @@
+/*
+ * Copyright (c) 2020, Renesas Electronics Corporation. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <stdint.h>
+#include <string.h>
+
+#include <common/debug.h>
+#include <lib/mmio.h>
+
+#include "cpg_registers.h"
+#include "rzg_def.h"
+#include "rzg_private.h"
+#include "rpc_registers.h"
+
+#define MSTPSR9_RPC_BIT			(0x00020000U)
+#define RPC_CMNCR_MD_BIT		(0x80000000U)
+#define RPC_PHYCNT_CAL			BIT(31)
+#define RPC_PHYCNT_STRTIM_G2M_V1	(0x6 << 15UL)
+#define RPC_PHYCNT_STRTIM		(0x7 << 15UL)
+
+static void rpc_enable(void)
+{
+	/* Enable clock supply to RPC. */
+	mstpcr_write(CPG_SMSTPCR9, CPG_MSTPSR9, MSTPSR9_RPC_BIT);
+}
+
+static void rpc_setup(void)
+{
+	uint32_t product, cut, reg, phy_strtim;
+
+	if (mmio_read_32(RPC_CMNCR) & RPC_CMNCR_MD_BIT)
+		mmio_clrbits_32(RPC_CMNCR, RPC_CMNCR_MD_BIT);
+
+	product = mmio_read_32(RZG_PRR) & PRR_PRODUCT_MASK;
+	cut = mmio_read_32(RZG_PRR) & PRR_CUT_MASK;
+
+	if ((product ==  PRR_PRODUCT_G2M) && (cut < PRR_PRODUCT_30))
+		phy_strtim = RPC_PHYCNT_STRTIM_G2M_V1;
+	else
+		phy_strtim = RPC_PHYCNT_STRTIM;
+
+	reg = mmio_read_32(RPC_PHYCNT);
+	reg &= ~RPC_PHYCNT_STRTIM;
+	reg |= phy_strtim;
+	mmio_write_32(RPC_PHYCNT, reg);
+	reg |= RPC_PHYCNT_CAL;
+	mmio_write_32(RPC_PHYCNT, reg);
+}
+
+void rzg_rpc_init(void)
+{
+	rpc_enable();
+	rpc_setup();
+}
diff --git a/drivers/renesas/rzg/rpc/rpc_registers.h b/drivers/renesas/rzg/rpc/rpc_registers.h
new file mode 100644
index 0000000..309abdb
--- /dev/null
+++ b/drivers/renesas/rzg/rpc/rpc_registers.h
@@ -0,0 +1,25 @@
+/*
+ * Copyright (c) 2020, Renesas Electronics Corporation. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#ifndef RPC_REGISTERS_H
+#define RPC_REGISTERS_H
+
+#define RPC_BASE	(0xEE200000U)
+#define RPC_CMNCR	(RPC_BASE + 0x0000U)
+#define RPC_SSLDR	(RPC_BASE + 0x0004U)
+#define RPC_DRCR	(RPC_BASE + 0x000CU)
+#define RPC_DRCMR	(RPC_BASE + 0x0010U)
+#define RPC_DRENR	(RPC_BASE + 0x001CU)
+#define RPC_SMCR	(RPC_BASE + 0x0020U)
+#define RPC_SMCMR	(RPC_BASE + 0x0024U)
+#define RPC_SMENR	(RPC_BASE + 0x0030U)
+#define RPC_CMNSR	(RPC_BASE + 0x0048U)
+#define RPC_DRDMCR	(RPC_BASE + 0x0058U)
+#define RPC_DRDRENR	(RPC_BASE + 0x005CU)
+#define RPC_PHYCNT	(RPC_BASE + 0x007CU)
+#define RPC_PHYINT	(RPC_BASE + 0x0088U)
+
+#endif /* RPC_REGISTERS_H */
-- 
2.7.4

